ExceptionHandling.md

```diff
@@ -0,0 +1,385 @@
+handle it.
+When using exceptions for control flow, programmers are forced to write catch blocks and the chance
+for swallowing an exception or decapitating an exception increases.
+
+Thus, avoid exceptions for expected or likely outcomes. For expected outcomes consider returning a
+meaningful result instead. (_Hint: Java Records present an easy way to implement result objects_)
+
+<!-- we use groogy as languate so the formatting works ;D -->
+
+```groovy
+public static boolean isValidProjectCode(String code) {
+   try {
+      new ProjectCode(code, "some value");
+   } catch (NullPointerException | IllegalArgumentException e) {
+      return false;
+   }
+   return true;
+}
+```
+
+becomes
+
+<!-- we use groogy as languate so the formatting works ;D -->
+
+```groovy
+public static boolean isValidProjectCode(String code) {
+   return ProjectCode.validate(code, "some value");
+}
+```
+
+When using custom exceptions as flow control, make them extend `Exception` and
+not `RuntimeException`
+because you intend the calling code to use them to determine the application flow.
+As previously discussed, all problems with checked exceptions [discussed here](#checked-exceptions)
+apply.
+I think this illustrates easily why exceptions for control flow can be a bad idea.
+
+If, on the other hand you reasonably expect a value to be present when no value is present, an
+exceptional state would follow.
+Thus throwing exceptions is fine in that case. For example:
+
+<!-- we use groogy as languate so the formatting works ;D -->
+
+```groovy
+public static void soSomeStuffWithMyProjectCode(String code) {
+   // code is always expected to be valid
+   var myProjectCode = new ProjectCode(code, "some value");
+   // ... (some complex logic involving myProjectCode)
+}
+
+public ProjectCode(String code, Constant constant) {
+   if (!validate(code, constant)) {
+      throw new RuntimeException("We encounter something unexpected.");
+   }
+}
+```
+
+In those cases we would not `catch` the exception and try
+to [handle it as late as possible](#throw-early-handle-late).
+
+**Exceptions as control-flow smell like a _GOTO_ instruction!**
+![](https://imgs.xkcd.com/comics/goto.png)
+
+## Good practices
+
+### Throw early, handle late
+
+When encountering exceptional circumstances, additional work should be avoided. Imagine an
+application calculating salaries for people working at a firm.
+One employee, Alice, left the company recently. Let's assume that the computation of a salary is
+complex.
+
+If the application would first calculate the salary for Alice and then try to retrieve her bank
+account from a list of employees,
+an exceptional situation would occur. _(Yes, one could argue that people leaving a company is not
+that uncommon but let's assume it is)_.
+
+Our application would fail after doing all the heavy lifting computing the salary. This is bad. A
+lot of resources wasted.
```

Comment:
NIT: 
```suggestion
Our application would fail after doing all the heavy lifting computing the salary. This is bad since a
lot of resources were wasted doing the salary computation.
```